### PR TITLE
Change HashDiff Constant to Hashdiff

### PR DESCRIPTION
**Pull request by jfelchner**
Why This Change Is Necessary
========================================================================

The name of this gem is `hashdiff`. There is another gem named
`hash_diff`. When they are both required via dependencies, errors are
thrown. Additionally since the behavior is different, gems that are
relying on `hashdiff`'s behavior, may get `hash_diff`'s behavior and
vice versa.

The reason for the conflict is because, based on ruby conventions, _'s
are the separators for camel case.

So:

```ruby
'hashdiff'.camelize # => "Hashdiff"

'hash_diff'.camelize # => "HashDiff"
```

Unfortunately instead of making the base module of this gem `Hashdiff`,
it's called `HashDiff` and it conflicts.

How These Changes Address the Issue
========================================================================

We change all of the constants to the new constant, including in
documentation and tests.

Then we add a shim to allow the old constant to be used and therefore
maintain backward compatibility.

Finally we add a warning message for users of the gem that the old
constant will be removed in the next major version bump.

Side Effects Caused By This Change
========================================================================

A new warning is added when loading the gem.

------------------------------------------------------------------------
References:
  * #45

### CODE REVIEW

**Comment by jfelchner** [comment]
Test failures seem to be Rubocop issues that existed before this PR.

**Comment by liufengyun** [comment]
Thanks a lot @jfelchner . WDYT @krzysiek1507 ?

File: lib/hashdiff.rb [hunk 1]
```diff
@@ -1,10 +1,14 @@
...
 
-require_relative 'hashdiff/util'
-require_relative 'hashdiff/compare_hashes'
-require_relative 'hashdiff/lcs'
-require_relative 'hashdiff/lcs_compare_arrays'
-require_relative 'hashdiff/linear_compare_array'
-require_relative 'hashdiff/diff'
-require_relative 'hashdiff/patch'
-require_relative 'hashdiff/version'
+require 'hashdiff/util'
```
Comment (krzysiek1507) [inline lib/hashdiff.rb]: Why did you change it from `require_relative` to `require`?

Comment (jfelchner) [reply lib/hashdiff.rb]: @krzysiek1507 it's in my commit message

Comment (krzysiek1507) [reply lib/hashdiff.rb]: @jfelchner could you please share more resources about this? Blog posts, issues etc.

Comment (jfelchner) [reply lib/hashdiff.rb]: Here's just one: https://github.com/rails/rails/issues/29846#issuecomment-338442315

Regardless there are many, many edge cases for using `require_relative` and it should never be used in a library.  Using it in a project is technically ok, although I never do that either.

Comment (luke-hill) [reply lib/hashdiff.rb]: Also the `/lib` path is added to the relative load directory, so these absolute paths will always work inside this gem (As they do inside most gems that are built in the normal way)

File: changelog.md [hunk 2]
```diff
@@ -22,7 +22,7 @@
 
 ## v0.3.4 2017-05-01
 
-* performance improvement of HashDiff#similar? #31
+* performance improvement of Hashdiff#similar? #31
```
Comment (luke-hill) [inline changelog.md]: I wouldn't change these 2 changelog messages, as your message from 0.4 onwards will be changing of the global scope.

File: lib/hashdiff.rb (hunk 1 above)
Comment (krzysiek1507) [reply lib/hashdiff.rb]: Ok, let's change it. I've never seen such issue before.

File: changelog.md (hunk 2 above)
Comment (krzysiek1507) [reply changelog.md]: I was thinking about it and I'm not sure about this change. @liufengyun what do you think?

Comment (luke-hill) [reply changelog.md]: Two options are ....
1) Revert change
2) (Preferred), use the syntax for `.` and `#` exclusively. So here this would say 

```
* performance improvement of `#similar?` #31
```

And the one below would say ...

```
add more specs for `.best_diff`
```

Comment (jfelchner) [reply changelog.md]: @luke-hill you're totally right.  Thanks for the catch!

Comment (jfelchner) [reply changelog.md]: Amended the commit.

**Comment by jfelchner** [comment]
@liufengyun just remember this needs to be version `0.4.0`.

**Comment by liufengyun** [comment]
Thank a lot @jfelchner 🎉 

And thanks for reviewing @krzysiek1507 @luke-hill !

**Comment by bquorning** [comment]
@liufengyun Could you please update the GitHub repository description as well?

image

**Comment by liufengyun** [comment]
@bquorning It's changed now, thanks for the catch.
